src/utils/hooks/useEmojiAutocomplete.ts

```diff
@@ -0,0 +1,215 @@
+  const containerRef = useOnBlur<ContainerElement>(() => setEmojiName(""));
+
+  const acceptSuggestion = useCallback(
+    (insertedEmoji: string) => {
+      if (cursor === null) return;
+
+      setValue((prev) => {
+        // end replace is cursor
+        // start replace is last /\s+/
+        const lastWordStart =
+          prev
+            .slice(0, cursor)
+            .split("")
+            .findLastIndex((c) => /\s+/.test(c)) + 1;
+        // remove the inputed :emoji_name add space behind emoji
+        return `${prev.slice(0, lastWordStart)}${insertedEmoji} ${prev.slice(cursor)}`;
+      });
+    },
+    [cursor]
+  );
+
+  // get suggestions for emoji name
+  useEffect(() => {
+    if (!emojiName) {
+      setSuggestions([]);
+      return;
+    }
+    // lazy load emoji data
+    if (!emojiData) {
+      import("constants/emojis.json").then((data) => setEmojiData(data.default as EmojiData[]));
+      return;
+    }
+
+    // update suggestions with filtered emojis
+    const matchedInName: EmojiData[] = [];
+    const matchedInSlug: EmojiData[] = [];
+
+    // no for ... of
+    emojiData.forEach((emoji) => {
+      const [slug, , , names] = emoji;
+      if (slug.includes(emojiName.toLowerCase())) matchedInSlug.push(emoji);
+      else if (names.some((name) => name.includes(emojiName.toLowerCase()))) matchedInName.push(emoji);
+    });
+
+    setSuggestions([...matchedInSlug, ...matchedInName]);
+  }, [emojiName, emojiData]);
+
+  // extract emoji name from value
+  useEffect(() => {
+    // reset emoji name and focused index when value changes
+    setEmojiName("");
+    setFocusedIndex(0);
+
+    if (cursor === null) return;
+
+    const lastWord = value.slice(0, cursor).split(/\s+/).pop();
+    if (!lastWord) return;
+
+    const [, newEmojiName] = lastWord.match(emojiRegex) || [];
+    if (!newEmojiName || newEmojiName.length < MIN_CHARACTERS_TO_TRIGGER_EMOJI_SUGGESTIONS) return;
+
+    if (lastWord.endsWith(":")) {
+      // find and insert emoji
+      const emoji = emojiData?.find(([slug]) => slug === newEmojiName);
+
+      if (emoji) {
+        const [, emojiString, supportsSkintones] = emoji;
+        acceptSuggestion(supportsSkintones ? emojiWithSkinTone(emojiString, skinToneComponent) : emojiString);
+      }
+
+      // dont set emoji name if emoji is not found
+      return;
+    }
+
+    // emoji autocomplete
+    setEmojiName(newEmojiName);
+  }, [value, cursor, emojiData, acceptSuggestion, skinToneComponent]);
+
+  // handle: update input, update suggestions
+  // FormEventHandler for TextareaAutosize
+  const handleChange: FormEventHandler<InputElement> & ChangeEventHandler<InputElement> = useCallback(
+    (e) => {
+      const newVal = e.currentTarget.value;
+      // prevent exceeding max input length by slicing the input
+      if (maxInputLength !== undefined && newVal.length > maxInputLength) setValue(newVal.slice(0, maxInputLength));
```

Comment:
```suggestion
      if (newVal.length > maxInputLength) setValue(newVal.slice(0, maxInputLength));
```

The check for undefined can be removed if you make the prop non-optional above

Reply:
the props are allready infered to allways be set, I wanted to make maxInputLength able to be `undefined` so that you can disable the input limit. It looks like I forgot to add `undefined` to the type